### PR TITLE
refactor: inspect/debugging as `useInspector`

### PR DESCRIPTION
**Pull request by threepointone**
This PR rewrites the logic for establishing a connection with the debugging socket, and adds some enhancements along the way.  Part of solving https://github.com/cloudflare/wrangler2/issues/188. Some highlights -

- I've installed `devtools-protocol`, a convenient package that has the static types for the devtools protocol (duh) autogenerated from chrome's devtools codebase.
- We now log messages and exceptions into the terminal directly, so you don't have to open devtools to see those messages.
- Messages are now buffered until a devtools instance connects, so you won't lose any messages while devtools isn't connected.
- We don't lose the connection on making changes to the worker, removing the need for the kludgy hack on the devtools side (where we refresh the whole page when there's a change)

Some things that I still have to do in followup PRs -
- clear the console whenever we make a change to the worker
- stay connected when we shift between local/remote mode

### CODE REVIEW

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [inline packages/wrangler/src/inspect.ts:502]: I still need to add logic for printing out these types, but I opened up the PR early anyway.

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [inline packages/wrangler/src/inspect.ts:129]: this is actually not needed, I should just use localWebSocket

File: packages/wrangler/src/inspect.ts
Comment (Electroid) [inline packages/wrangler/src/inspect.ts:267]: What's broken about it?

File: packages/wrangler/src/inspect.ts
Comment (Electroid) [reply packages/wrangler/src/inspect.ts:502]: Is there a library we can use?

Comment (threepointone) [reply packages/wrangler/src/inspect.ts:502]: couldn't find any

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [reply packages/wrangler/src/inspect.ts:267]: Just doesn't enable the Network panel in exisiting devtools, works on our forked version.

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [reply packages/wrangler/src/inspect.ts:129]: I'm mistaken, javascript scopes mean I have to keep this boolean around.

File: packages/wrangler/src/dev.tsx
Comment (threepointone) [inline packages/wrangler/src/dev.tsx:167]: in a followup PR, we'll make this port configurable

File: packages/wrangler/src/inspect.ts
Comment (JacobMGEvans) [inline packages/wrangler/src/inspect.ts:284]: Is there an upper limit on this, or retries indefinitely?

Comment (threepointone) [reply packages/wrangler/src/inspect.ts:284]: Indefinitely. which is... fine, I think. It takes a minute or so between retries, and stops the moment you open the browser tab or devtools. We can revisit this if it's a problem.

File: packages/wrangler/src/inspect.ts
Comment (petebacondarwin) [inline packages/wrangler/src/inspect.ts:336]: Why not just buffer all messages, since if the socket was active we would have forwarded these messages on anyway?

File: packages/wrangler/src/inspect.ts
Comment (petebacondarwin) [inline packages/wrangler/src/inspect.ts:388]: And if we renamed `remoteMessageListener` to `sendMessageToLocal` then it would have been more obvious to reuse that here:

```suggestion
      messageBufferRef.current.forEach(sendMessageToLocal);
```

File: packages/wrangler/src/inspect.ts
Comment (petebacondarwin) [inline packages/wrangler/src/inspect.ts:419]:
```suggestion
 * into arguments suitable to be called by a console method, and
```

File: packages/wrangler/src/inspect.ts
Comment (petebacondarwin) [inline packages/wrangler/src/inspect.ts:525]: Given that this is running in ink, I wonder if we could pass these messages out to be rendered in an Ink component, rather than just dumping them out using `console.xxx`?
(I appreciate that Ink [patches `console`](https://github.com/vadimdemedes/ink#patchconsole) but I think we could be more elegant here...

File: packages/wrangler/src/inspect.ts
Comment (petebacondarwin) [inline packages/wrangler/src/inspect.ts:130]: `console.warn` or `console.error` here?

File: packages/wrangler/src/inspect.ts
Comment (petebacondarwin) [inline packages/wrangler/src/inspect.ts:278]: Is this just something that you have noticed or is it documented somewhere?

**Review by petebacondarwin — APPROVED** [review]
I love the use of hooks here to keep dependencies in sync. Very nice!

I made a bunch of comments, but I don't think any are blocking this PR, assuming that you have actually checked it works as expected!

Can you provide some thoughts on how you propose we write tests for all this functionality?

File: packages/wrangler/src/inspect.ts
Comment (JacobMGEvans) [reply packages/wrangler/src/inspect.ts:170]: Signal?

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [reply packages/wrangler/src/inspect.ts]: Huh, I could swear I tested this. Thanks for the catch!

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [reply packages/wrangler/src/inspect.ts]: nice. will do.

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [reply packages/wrangler/src/inspect.ts:168]: I want a good word for *something* that's used to trigger the effect again. A sigil is used through history as a sign/seal of something. I would say counter, but that's the implementation detail. Got any suggestions for something better?

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [reply packages/wrangler/src/inspect.ts:355]: The message is brittle. I highly dount it's part of the spec. Honestly maybe this can just be silent. When I add a logger for https://github.com/cloudflare/wrangler2/issues/230 I'll revisit this. Leaving this conversation unresolved so I can find it later.

File: packages/wrangler/src/inspect.ts
Comment (threepointone) [reply packages/wrangler/src/inspect.ts:523]: We can revisit it. I actually try to avoid Ink for anything that doesn't feel like UI (and before you say it, I understand the boudnary is deeply fuzzy.) For example, in this usecase, I want the messages to persist no matter what, even if the process exits, etc. For now I think calling `console` methods makes sense, and matches behaviour closer to node (or at least, is already better than what we have, which is nothing.)

**Review by threepointone — COMMENTED** [review]
I updated the PR with suggested changes. I'm going to write down some followup tasks, including testing plan, before I land this. Thanks @petebacondarwin!

**Pull request merged**
